### ISSUE TITLE
syncGridSelection and preserveHidden

### ISSUE THREAD

**Issue by kchristoudias** [ghiscoding/Angular-Slickgrid#191]
Is there any way I can set the preserveHidden option? I couldn't find any way to do that. 
Ref: https://github.com/mleibman/SlickGrid/wiki/DataView ->  Synchronizing selection & cell CSS styles

**Comment by ghiscoding** [comment]
Never heard about that flag before and looking at the code, I'm not sure how to even implement this. 
I would suggest you to use directly the SlickGrid grid & dataView objects, see this [Wiki](https://github.com/ghiscoding/Angular-Slickgrid/wiki/SlickGrid-&-DataView-Objects). 
Or you could always create a PR to implement the feature where necessary.

Question to you.. is there a particular reason you didn't you use the issue template? These templates helps to provide a better picture of user's problem, but you're the 2nd person that post a question without using the template

**Comment by kchristoudias** [comment]
Okay I will investigate more and if I find anything I will reply here. I am not sure how to do this either, I needed this because when selecting rows with checkboxes, and filtering data, the selections are lost during the process and I found out that in the javascript project you can use the syncGridSelection function with the second parameter true (preserveHidden) and fix this.

I tried copying the library locally and changing the preserveHidden value to be always true and it seems to be working. But I will try to use the dataView directly as you suggested and let you know. Thank you!!

Sorry for not using the template, first time i used the new issue button it just took me to the regular issue straight away and didnt even see the templates. I dont know why though because i tried again now and I can see the templates normally.

**Comment by ghiscoding** [comment]
I would not mind receiving a PR with this feature to incorporate in the library, I just don't know where I would put that and I also don't understand it well. Would we set this up during grid creation? Or would it be ok to add this to the GridService? I'm not sure about the best approach and where to add it, so if you find the good way of doing it and want to make a PR, I'm all for it. 

For the issue template, I was actually wondering if it comes up all the time, if it doesn't then that might explain why. 

Thanks for the feedback

**Comment by kchristoudias** [comment]
After trying different methods I found a way to make this work. 

My code before

```  
 this.gridOptions = {
      enableFiltering: true,
      enableGrouping : true,
      enableCheckboxSelector: true,
      dataView : {
        syncGridSelection : true
      }
    };
    this.dataset = [];
  }

  angularGridReady(angularGrid: AngularGridInstance) {
    this.angularGrid = angularGrid;
    this.gridObj = angularGrid && angularGrid.slickGrid || {};
    this.dataViewObj = angularGrid.dataView;

  }
```

And my code after ( working )

```   
 this.gridOptions = {
      enableFiltering: true,
      enableGrouping : true,
      enableCheckboxSelector: true
    };
    this.dataset = [];
  }

  angularGridReady(angularGrid: AngularGridInstance) {
    this.angularGrid = angularGrid;
    this.gridObj = angularGrid && angularGrid.slickGrid || {};
    this.dataViewObj = angularGrid.dataView;

   this.dataViewObj.syncGridSelection(this.gridObj, true , true);
  }
 ```

I am not really sure why the syncGridSelection in gridOptions creates this problem but by removing it and setting/calling the function through the dataView object in the onAngularGridCreated event seems to fix this problem. I hope this will help someone in the future

**Comment by ghiscoding** [comment]
@kchristoudias 
I actually forgot that I had set this property in the Angular-Slickgrid component, which you can see on this [line](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L286) which might explain why it doesn't work. After finding this [commit](https://github.com/ghiscoding/Angular-Slickgrid/commit/1c3c163b58540a17e1abeccf0febb048830a7164#diff-6011989ccee0745f4b7a9f44753c6036R267) of when this was set up in the code, it seems to be related to the Frozen Columns feature, I slightly recall that it was causing some issue with that feature. 

From what I can see, I never used the 3rd argument like you did, only 2 arguments is set in the code, so by calling it that way in Angular-Slickgrid itself, the 3rd argument is probably defaults to false while you want it to be true.  This 3rd argument seem to have been added later and is not even in the DataView [Wiki](https://github.com/6pac/SlickGrid/wiki/DataView#synchronizing-selection--cell-css-styles), so the only doc is directly in the code on this [line](https://github.com/6pac/SlickGrid/blob/master/slick.dataview.js#L988) of code

To resolve this issue, I think we can change the GridOption Interface [here](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/models/gridOption.interface.ts#L105) with the following changes
```diff
  dataView?: {
    /**
     * if you don't want the items that are not visible (due to being filtered out
     * or being on a different page) to stay selected, then set this property as 'false'
     */
-    syncGridSelection?: boolean; 
+    syncGridSelection?: boolean | { preserveHidden: boolean; preserveHiddenOnSelectionChange: boolean; };
  };
```
and in Angular-Slickgrid Component [here](https://github.com/ghiscoding/Angular-Slickgrid/blob/master/src/app/modules/angular-slickgrid/components/angular-slickgrid.component.ts#L286), we could change to 
```diff
// if you don't want the items that are not visible (due to being filtered out
// or being on a different page) to stay selected, pass 'false' to the second arg
if (this.gridOptions && this.gridOptions.dataView && this.gridOptions.dataView.hasOwnProperty('syncGridSelection')) {
+  const syncGridSelection = this.gridOptions.dataView.syncGridSelection;
+  if (typeof syncGridSelection === 'boolean') {
    this.dataView.syncGridSelection(this.grid, this.gridOptions.dataView.syncGridSelection);
+  } else {
+    this.dataView.syncGridSelection(this.grid, syncGridSelection.preserveHidden, syncGridSelection.preserveHiddenOnSelectionChange);
+  }
}
```
By making the option a boolean or an object with the 2 properties, this bring the benefit of not breaking anyone's code and also add a way to change these 2 flags. 

Thoughts?

**Comment by kchristoudias** [comment]
@ghiscoding

Yes I believe this would definitely work! Actually while I was debugging to find a solution for this, I could see that the 3rd parameter was undefined ( not false ) and the second argument was true. 

Thus making the change you are suggesting would be the best so people with the Frozen Columns feature will not have problems with their code breaking and actually making this parametric for anyone in the future to use.

**Comment by ghiscoding** [comment]
Note that I'm waiting for a new version release of SlickGrid core lib before releasing a new version on my side. I'll write back when this is released

**Comment by ghiscoding** [comment]
Now released under version [2.7.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.7.0)

**Comment by kchristoudias** [comment]
of course I will vote! thank you for resolving this!
Cheers!!
